File: Binary_Search_Tree/538.Convert_BST_to_Greater_Tree/Python/538.Convert_BST_to_Greater_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def convertBST(self, root):
        """
        :type root: Optional[TreeNode]
        :rtype: Optional[TreeNode]
        """
        #======================================#
        # Recursion-based DFS traversal method #
        #======================================#
        if (not root): #Issue/Boundary-case handling
            return None

        ############
        #Initialize
        ##### Record summary value #####
        record_sum_val = [0]

        ####################################
        #Recursion-based DFS loop traversal
        self.dfsTraversal(root, record_sum_val) #Recursion function call (i.e. whole root tree-node)

        return root

    
    def dfsTraversal(self, root, record_sum_val):
        """
        :type root: Optional[TreeNode]
        :type record_sum_val: List[int]
        :rtype: None, void
        """
        #========================================#
        # Recursion-based DFS traversal function #
        #========================================#
        if (not root): #Issue/Boundary-case handling
            return

        ####################
        #Whole process/flow
        self.dfsTraversal(root.right, record_sum_val) #Recursion function call (i.e. root right-half node)

        record_sum_val[0] += root.val #Update/Accumulate

        root.val = record_sum_val[0] #Update/Overwrite

        self.dfsTraversal(root.left, record_sum_val) #Recursion function call (i.e. root left-half node)

File: Binary_Search_Tree/538.Convert_BST_to_Greater_Tree/C++/538.Convert_BST_to_Greater_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode * convertBST(TreeNode * root) 
    {
        //========================================//
        // Input type:                            //
        //   - TreeNode * (root tree-node)        //
        // Return type:                           //
        //   - TreeNode * (result root tree-node) //
        //========================================//
        if (root == nullptr) { return nullptr; } //Issue, Boundary-case handling

        /*Initialize*/
        ///// Record summary value /////
        int record_sum_val = 0;

        /*Recursion-based DFS loop traversal*/
        dfsTraversal(root, record_sum_val); //Recursion function call (i.e. whole root tree-node)

        return root;
    }


    void dfsTraversal(TreeNode * root, int & record_sum_val)
    {
        //=====================================//
        // Input type:                         //
        //   - TreeNode * (root tree-node)     //
        //   - int ref. (record summary value) //
        // Return type:                        //
        //   - void (no return)                //
        //=====================================//
        if (root == nullptr) { return; } //Issue, Boundary-case handling

        /*Whole process, flow*/
        dfsTraversal((root->right), record_sum_val); //Recursion function call (i.e. root right-half node)

        record_sum_val += (root->val); //Update, Accumulate

        (root->val) = record_sum_val; //Update, Overwrite

        dfsTraversal((root->left), record_sum_val); //Recursion function call (i.e. root left-half node)
    }
};
